lib/tests.js

```diff
@@ -0,0 +1,361 @@
+
+
+/**
+ * Creates a test object.
+ *
+ * @param {assert} assert
+ * @param {Object} config
+ * @param {Object} Plugin DDQ instance being tested.
+ * @return {ddqBackendTests.test}
+ */
+module.exports = (assert, config, Plugin) => {
+    var instance;
+
+    instance = new Plugin(config);
+
+
+    /**
+     * Runs polling test.
+     *
+     * @param {Function} callback
+     */
+    function pollingTest(callback) {
+        var instanceProperties;
+
+        instanceProperties = [
+            "currentlyPolling",
+            "currentlyRestoring",
+            "poller",
+            "restorer"
+        ];
+
+
+        /**
+         * Assert properties in batches.
+         *
+         * @param {string[]} properties
+         * @param {*} [shouldExist] asserts the properties exist or does not exist.
+         */
+        function assertProperties(properties, shouldExist) {
+            properties.forEach((property) => {
+                if (shouldExist) {
+                    assert(instance[property]);
+                } else {
+                    assert(!instance[property]);
+                }
+            });
+        }
+
+
+        /**
+         * Starts and stops listening while asserting properties.
+         *
+         * @param {Function} done
+         */
+        function startStopAssert(done) {
+            instance.startListening(() => {
+                assertProperties(instanceProperties, true);
+                instance.stopListening(() => {
+                    assertProperties(instanceProperties);
+                    done();
+                });
+            });
+        }
+
+        assertProperties(instanceProperties);
+        instance.connect((connectErr) => {
+            if (connectErr) {
+                console.log("There was a connection error");
+                callback(connectErr);
+
+                return;
+            }
+
+            // Assert properties twice to ensure it works properly.
+            startStopAssert(() => {
+                startStopAssert(() => {
+                    callback();
+                });
+            });
+        });
+    }
+
+
+    /**
+     * Runs test on particular message. Blows up after 5 tries.
+     * Sets up listener and calls callback when the necessary message is
+     * emitted.
+     *
+     * @param {Function} callback
+     */
+    function runOnMessage(callback) {
+        var runCounter;
+
+        runCounter = 0;
+
+        instance.on("data", (data) => {
+            instance.stopListening();
+
+            if (data.message === "message") {
+                callback(null, data);
+            } else {
+                runCounter += 1;
+
+                if (runCounter > 5) {
+                    callback(new Error("The test message was not found."));
+                }
+            }
+        });
+
+        instance.startListening();
+    }
+
+
+    /**
+     * Requeue test.
+     *
+     * @param {Function} callback
+     */
+    function requeue(callback) {
```

Comment:
We discussed this sort of. It might just be me but renaming these may be beneficial. I keep getting them confused in my mind with the instances functions. This doesn't just run requeue but sets up a listener, waits to get the message and then requeues it.